### PR TITLE
feat(oracledb)!: Update to SDK V3 Arrow native

### PR DESCRIPTION
**Pull request by erezrokah**
Closes https://github.com/cloudquery/cloudquery/issues/10760

BEGIN_COMMIT_OVERRIDE
feat: Update to use [Apache Arrow](https://arrow.apache.org/) type system (#11115)

BREAKING-CHANGE: This release introduces an internal change to our type system to use [Apache Arrow](https://arrow.apache.org/). This should not have any visible breaking changes, however due to the size of the change we are introducing it under a major version bump to communicate that it might have some bugs that we weren't able to catch during our internal tests. If you encounter an issue during the upgrade, please submit a [bug report](https://github.com/cloudquery/cloudquery/issues/new/choose). You will also need to update destinations depending on which one you use:
- Azure Blob Storage >= v3.2.0
- BigQuery >= v3.0.0
- ClickHouse >= v3.1.1
- DuckDB >= v1.1.6
- Elasticsearch >= v2.0.0
- File >= v3.2.0
- Firehose >= v2.0.2
- GCS >= v3.2.0
- Gremlin >= v2.1.10
- Kafka >= v3.0.1
- Meilisearch >= v2.0.1
- Microsoft SQL Server >= v4.2.0
- MongoDB >= v2.0.1
- MySQL >= v2.0.2
- Neo4j >= v3.0.0
- PostgreSQL >= v4.2.0
- S3 >= v4.4.0
- Snowflake >= v2.1.1
- SQLite >= v2.2.0

END_COMMIT_OVERRIDE

### CODE REVIEW

File: plugins/source/oracledb/client/types.go [hunk 1]
```diff
@@ -3,59 +3,50 @@ package client
...
-		schema.TypeInet:
-		return "clob"
 	default:
-		panic("unknown type " + t.String())
+		return "clob"
 	}
 }
 
-func SchemaType(tableName string, columnName string, dataType string) schema.ValueType {
-	if strings.HasPrefix(columnName, "timestamp") {
```
Comment (erezrokah) [inline plugins/source/oracledb/client/types.go:41]: There was a bug here actually. We should check `dataType` and not `columnName`.

Also we can remove the `tableName` and `columnName` arguments

File: plugins/source/oracledb/client/sync.go [hunk 2]
```diff
@@ -27,22 +29,23 @@ func (*Client) createResultsArray(table *schema.Table) []any {
 	results := make([]any, 0, len(table.Columns))
 	for _, col := range table.Columns {
 		switch col.Type {
-		case schema.TypeUUID, schema.TypeByteArray:
+		case types.ExtensionTypes.UUID, arrow.BinaryTypes.Binary, arrow.BinaryTypes.LargeBinary:
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/sync.go:32]: fixed size binary?

File: plugins/source/oracledb/client/sync.go [hunk 3]
```diff
@@ -27,22 +29,23 @@ func (*Client) createResultsArray(table *schema.Table) []any {
...
+		case arrow.PrimitiveTypes.Int8, arrow.PrimitiveTypes.Int16, arrow.PrimitiveTypes.Int32, arrow.PrimitiveTypes.Int64,
+			arrow.PrimitiveTypes.Uint8, arrow.PrimitiveTypes.Uint16, arrow.PrimitiveTypes.Uint32, arrow.PrimitiveTypes.Uint64:
 			var r *int
 			results = append(results, &r)
-		case schema.TypeFloat:
+		case arrow.FixedWidthTypes.Float16, arrow.PrimitiveTypes.Float32, arrow.PrimitiveTypes.Float64:
 			var r *float64
 			results = append(results, &r)
-		case schema.TypeTimestamp:
+		case arrow.FixedWidthTypes.Timestamp_us:
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/sync.go:45]:
```suggestion
		case *arrow.TimestampType:
```

File: plugins/source/oracledb/client/transformer.go [hunk 4]
```diff
@@ -1,137 +1,82 @@
...
+		return a.Value(i), nil
+	case *array.Uint8:
+		return a.Value(i), nil
+	case *array.Uint16:
+		return a.Value(i), nil
+	case *array.Uint32:
+		return a.Value(i), nil
+	case *array.Uint64:
+		return a.Value(i), nil
+	case *array.Float32:
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/transformer.go:32]: float16? (you mentioned it in `sync.go` ([`float16.Num`](https://pkg.go.dev/github.com/apache/arrow/go/arrow/float16#Num) is returned & you can use [`float16.Num.Float32`](https://pkg.go.dev/github.com/apache/arrow/go/arrow/float16#Num.Float32) method)

File: plugins/source/oracledb/client/transformer.go [hunk 5]
```diff
@@ -1,137 +1,82 @@
...
+		return a.Value(i), nil
+	case *array.Float64:
+		return a.Value(i), nil
+	case *array.String:
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/transformer.go:36]: LargeString?

IMO, the better way is to check for `array.BinaryLike` and then use `IsUtf8` to verify whether it's string or byte array

File: plugins/source/oracledb/client/transformer.go [hunk 6]
```diff
@@ -1,137 +1,82 @@
...
+		switch ts.DataType().(*arrow.TimestampType).Unit {
+		case arrow.Nanosecond:
+			return ts.Value(i).ToTime(arrow.Nanosecond), nil
+		case arrow.Millisecond:
+			return ts.Value(i).ToTime(arrow.Millisecond), nil
+		case arrow.Second:
+			return ts.Value(i).ToTime(arrow.Second), nil
+		default:
+			return ts.Value(i).ToTime(arrow.Microsecond), nil
+		}
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/transformer.go:55]:
```suggestion
		toTime, err := a.DataType().(*arrow.TimestampType).GetToTimeFunc()
		if err != nil {
			return nil, err
		}
		return toTime(a.Value(i)), nil
```

File: plugins/source/oracledb/client/transformer.go [hunk 7]
```diff
@@ -1,137 +1,82 @@
...
+		for j := 0; int64(j) < record.NumCols(); j++ {
+			v, err := GetValue(record.Column(j), int(i))
+			if err != nil {
+				return nil, err
+			}
+			row = append(row, v)
+		}
+		res = append(res, row)
+	}
+	return res, nil
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/transformer.go:81]: may prealloc, nit
```suggestion
	res := make([][]any, record.NumRows())
	var err error
	for i := int64(0); i < record.NumRows(); i++ {
		row := make([]any, record.NumCols())
		for j := 0; int64(j) < record.NumCols(); j++ {
			row[j], err = GetValue(record.Column(j), int(i))
			if err != nil {
				return nil, err
			}
		}
		res[i] = row
	}
	return res, nil
```

File: plugins/source/oracledb/client/types.go [hunk 8]
```diff
@@ -3,59 +3,50 @@ package client
...
+		*arrow.Uint8Type, *arrow.Uint16Type, *arrow.Uint32Type, *arrow.Uint64Type:
 		return "NUMBER(19)"
-	case schema.TypeFloat:
+	case *arrow.Float16Type, *arrow.Float32Type, *arrow.Float64Type:
 		return "binary_double"
-	case schema.TypeUUID:
+	case *types.UUIDType:
 		return "raw(16)"
-	case schema.TypeByteArray:
+	case *arrow.BinaryType, *arrow.LargeBinaryType:
```
Comment (candiduslynx) [inline plugins/source/oracledb/client/types.go:21]: fixed size binary?

File: plugins/source/oracledb/resources/plugin/plugin_test.go [hunk 9]
```diff
@@ -150,12 +163,12 @@ func TestPlugin(t *testing.T) {
 	if err := createTable(ctx, db, testTable); err != nil {
 		t.Fatal(err)
 	}
-	data := testdata.GenTestDataV1(testTable)
+	data := schema.GenTestData(testTable, schema.GenTestDataOptions{MaxRows: 1})
```
Comment (candiduslynx) [inline plugins/source/oracledb/resources/plugin/plugin_test.go:166]: what about null row (available if we gen 2 rows)?

File: plugins/source/oracledb/client/sync.go (hunk 2 above)
Comment (erezrokah) [reply plugins/source/oracledb/client/sync.go]: Cleaned this up in https://github.com/cloudquery/cloudquery/pull/11115/commits/18f83f179999630a6272f36892a9317d50651bcd as we only need to support the arrows types we use based on the DB schema

File: plugins/source/oracledb/client/transformer.go (hunk 5 above)
Comment (erezrokah) [reply plugins/source/oracledb/client/transformer.go:38]: Done in https://github.com/cloudquery/cloudquery/pull/11115/commits/e034e620676f8f91cee5e04c15241c7c11cb7bc2. Not sure I follow the `array.BinaryLike` comment

File: plugins/source/oracledb/client/transformer.go (hunk 4 above)
Comment (erezrokah) [reply plugins/source/oracledb/client/transformer.go:34]: Done in https://github.com/cloudquery/cloudquery/pull/11115/commits/e034e620676f8f91cee5e04c15241c7c11cb7bc2

File: plugins/source/oracledb/client/transformer.go (hunk 6 above)
Comment (erezrokah) [reply plugins/source/oracledb/client/transformer.go]: Done in https://github.com/cloudquery/cloudquery/pull/11115/commits/e034e620676f8f91cee5e04c15241c7c11cb7bc2

File: plugins/source/oracledb/client/types.go (hunk 8 above)
Comment (erezrokah) [reply plugins/source/oracledb/client/types.go]: Done in https://github.com/cloudquery/cloudquery/pull/11115/commits/380cb2908119335e9c983cf4272492244b8bfe7f

File: plugins/source/oracledb/client/transformer.go (hunk 5 above)
Comment (candiduslynx) [reply plugins/source/oracledb/client/transformer.go:38]: @erezrokah 
For the `array.BinaryLike`:
```go
case array.BinaryLike:
	if a.DataType().(arrow.BinaryDataType).IsUtf8() {
		return arrow.BinaryTypes.String
	}
	return arrow.BinaryTypes.Binary
```

File: plugins/source/oracledb/resources/plugin/plugin_test.go (hunk 9 above)
Comment (erezrokah) [reply plugins/source/oracledb/resources/plugin/plugin_test.go:166]: Tried it and currently blocked by https://github.com/cloudquery/plugin-sdk/pull/922

Comment (erezrokah) [reply plugins/source/oracledb/resources/plugin/plugin_test.go]: https://github.com/cloudquery/cloudquery/pull/11115/commits/c9a80fd657ec924ebf1f9380c1874f12137d03b0 should test for null values and also properly support `number` schema type using `decimal`

File: plugins/source/oracledb/resources/plugin/plugin_test.go [hunk 10]
```diff
@@ -52,20 +53,24 @@ func createTable(ctx context.Context, db *sql.DB, table *schema.Table) error {
...
+			// In OracleDB primary keys are implicitly NOT NULL and UNIQUE
+			// and it errors out if we try to do it explicitly
+			if column.NotNull {
+				builder.WriteString(" NOT NULL")
+			}
+			if column.Unique {
+				builder.WriteString(" UNIQUE")
+			}
 		}
+
```
Comment (candiduslynx) [inline plugins/source/oracledb/resources/plugin/plugin_test.go:73]: `continue` early for less nesting, nit
```suggestion
			continue
		}

		// In OracleDB primary keys are implicitly NOT NULL and UNIQUE
		// and it errors out if we try to do it explicitly
		if column.NotNull {
			builder.WriteString(" NOT NULL")
		}
		if column.Unique {
			builder.WriteString(" UNIQUE")
		}
```

File: plugins/source/oracledb/resources/plugin/plugin_test.go [hunk 11]
```diff
@@ -52,20 +53,24 @@ func createTable(ctx context.Context, db *sql.DB, table *schema.Table) error {
...
 		if i < len(table.Columns)-1 {
```
Comment (candiduslynx) [inline plugins/source/oracledb/resources/plugin/plugin_test.go:74]: can be moved to the beginning of the loop with `if i > 0`, nit

**Review by candiduslynx — APPROVED** [review]
just a couple of nits, LGTM

File: plugins/source/oracledb/resources/plugin/plugin_test.go [hunk 12]
```diff
@@ -118,6 +131,75 @@ func isNotExistsError(err error) bool {
...
 }
 
+func getActualStringsForResource(t *testing.T, resource *schema.Resource) []string {
+	actualData := resource.GetValues()
+	actualStrings := make([]string, len(actualData))
+	for i, v := range actualData {
+		// Oracle treats empty strings as null, so we need to convert them for comparison
+		// See https://stackoverflow.com/questions/203493/why-does-oracle-9i-treat-an-empty-string-as-null
+		// Please note this means we can't distinguish between null and empty string
+		if !v.IsValid() && v.DataType() == arrow.BinaryTypes.String {
```
Comment (candiduslynx) [inline plugins/source/oracledb/resources/plugin/plugin_test.go:141]: Use arrow.TypeEqual

**Review by candiduslynx — APPROVED** [review]

**Review by candiduslynx — APPROVED** [review]

**Comment by erezrokah** [comment]
This is on hold until for the same reason as https://github.com/cloudquery/cloudquery/pull/11212 (an issue with CQ internal columns in SDK v3)

**Comment by erezrokah** [comment]
Going to merge this as https://github.com/cloudquery/cloudquery/pull/11214#issuecomment-1600768285

**Pull request merged**